### PR TITLE
Discrete token-based TTS implementation

### PR DESCRIPTION
**Pull request by ftshijt**
## What?

The new task of discrete token-based TTS implementation (mostly focuses on acoustic modeling, by changing spectrogram to discrete token as output prediction)

## TODO

- [ ] Integrate monotonic attention implementation (verified but doing code cleaning)
- [x] Add more CI

### CODE REVIEW

**Comment by ftshijt** [comment]
Hi @sw005320 , may I kindly request your review on this PR? It would be great to merge into master as a reference of discrete TTS (also we could accelerate this direction with some of our recent plans for the codec design)

I'm thinking about adding the monotonic attention implementation (originally planed in this PR) in another PR cause it was not related to the task design and the current PR is already a large one

**Comment by sw005320** [comment]
Sorry, it was not under my radar, as the TODO items are not marked.
I'll start to review this.

**Comment by sw005320** [comment]
@simpleoier, can you help review this PR?

**Comment by sw005320** [comment]
@kan-bayashi, can you also review this PR?
Since this PR touches the core algorithm part, I want your help.

**Review by sw005320 — COMMENTED** [review]
Hi @ftshijt, I want to leave the review of the template by @simpleoier and the model parts by @kan-bayashi.

My major comment is that we need to add some unit tests.

File: egs2/ljspeech/tts1/conf/tuning/train_tacotron2.yaml
[the remark this answers is not in the document]
Comment (ftshijt) [reply egs2/ljspeech/tts1/conf/tuning/train_tacotron2.yaml:1]: This is not intended, let me revert it.

File: espnet2/bin/tts2_inference.py [hunk 1]
```diff
@@ -0,0 +1,730 @@
...
+
+import numpy as np
+import soundfile as sf
+import torch
+from packaging.version import parse as V
+from typeguard import check_argument_types
+
+from espnet2.fileio.npy_scp import NpyScpWriter
+
+# from espnet2.gan_tts.vits import VITS
```
Comment (simpleoier) [inline espnet2/bin/tts2_inference.py:21]: remove this line? (please check the following as well).

File: espnet2/tasks/tts2.py [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+import logging
+from pathlib import Path
+from typing import Callable, Collection, Dict, List, Optional, Tuple, Union
+
+import numpy as np
+import torch
+import yaml
+from typeguard import check_argument_types, check_return_type
+
+# from espnet2.gan_tts.jets import JETS
```
Comment (simpleoier) [inline espnet2/tasks/tts2.py:13]: remove these unused lines? (also check the following)

File: espnet2/tasks/tts2.py [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+# from espnet2.tts.prodiff import ProDiff
+# from espnet2.tts.tacotron2 import Tacotron2
+# from espnet2.tts.transformer import Transformer
+from espnet2.tts.utils import ParallelWaveGANPretrainedVocoder
+from espnet2.utils.get_default_kwargs import get_default_kwargs
+from espnet2.utils.griffin_lim import Spectrogram2Waveform
+from espnet2.utils.nested_dict_action import NestedDictAction
+from espnet2.utils.types import int_or_none, str2bool, str_or_none
+
+discrete_feats_extractor_choices = ClassChoices(
```
Comment (simpleoier) [inline espnet2/tasks/tts2.py:49]: How about importing these from tts to remove redundancy?

File: espnet2/tts2/fastspeech2/fastspeech2_discrete.py
Comment (simpleoier) [inline espnet2/tts2/fastspeech2/fastspeech2_discrete.py:1]: Do you think if it is better to inherit from `fastspeech2.py`?

File: egs2/TEMPLATE/tts2/tts2.sh [hunk 4]
```diff
@@ -0,0 +1,1283 @@
...
+storage_save_mode=true          # Save storage on SSL feature extraction
+                                # If true, feature extraction and kmeans clustering on the fly
+
+
+# X-Vector related
+use_xvector=false   # Whether to use x-vector.
+xvector_tool=kaldi  # Toolkit for extracting x-vector (speechbrain, rawnet, espnet, kaldi)
+xvector_model=speechbrain/spkrec-ecapa-voxceleb  # For only espnet, speechbrain, or rawnet
+
+# Vocabulary related
```
Comment (simpleoier) [inline egs2/TEMPLATE/tts2/tts2.sh:89]: Are oov, blank, sos_eos necessary?

File: egs2/TEMPLATE/tts2/tts2.sh [hunk 5]
```diff
@@ -0,0 +1,1283 @@
...
+              --input "${data_feats}/srctexts" --output "${src_token_list}" \
+              --non_linguistic_symbols "${nlsyms_txt}" \
+              --cleaner "${cleaner}" \
+              --g2p "${g2p}" \
+              --write_vocabulary true \
+              --add_symbol "${blank}:0" \
+              --add_symbol "${oov}:1" \
+              --add_symbol "${sos_eos}:-1"
+
+        # Jinchuan: also build the tgt_vocab even though this is quite naive.
```
Comment (simpleoier) [inline egs2/TEMPLATE/tts2/tts2.sh:568]: Maybe using Note(Jinchuan) is better.

File: egs2/TEMPLATE/tts2/tts2.sh [hunk 6]
```diff
@@ -0,0 +1,1283 @@
...
+                _opts+=" --option ${dumpdir}/xvector/${dset}/spk_xvector.ark"
+            done
+        fi
+        if "${use_sid}"; then
+            _opts+=" --option ${data_feats}/org/${train_set}/spk2sid"
+        fi
+        if "${use_lid}"; then
+            _opts+=" --option ${data_feats}/org/${train_set}/lang2lid"
+        fi
+        ${python} -m espnet2.bin.pack tts2 \
```
Comment (simpleoier) [inline egs2/TEMPLATE/tts2/tts2.sh:1160]: Is the kmeans model packed here?

File: egs2/TEMPLATE/tts2/tts2.sh (hunk 4 above)
Comment (ftshijt) [reply egs2/TEMPLATE/tts2/tts2.sh:89]: Yeah, this is used in the text tokenizer. Though blank is not necessary for TTS, we still use it to align with ASR/TTS side of the tokenizer.

File: espnet2/tts2/fastspeech2/fastspeech2_discrete.py
Comment (ftshijt) [reply espnet2/tts2/fastspeech2/fastspeech2_discrete.py:1]: I would prefer to separate in the setting.

One reason is that the fatspeech2 one is quite established (likely no updates in the future), but this discrete one could have several potential directions to update later. The other is that some following models in tts2 would be totally different from the TTS setting, which is likely to result in inconsistent inheritance.

File: egs2/TEMPLATE/tts2/tts2.sh (hunk 6 above)
Comment (ftshijt) [reply egs2/TEMPLATE/tts2/tts2.sh:1163]: Fixed

File: egs2/ljspeech/tts2/run.sh [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+./tts2.sh \
+    --nj 16 \
+    --inference_nj 16 \
+    --fs 16000 --n_shift 320 --n_fft 1280 \
+    --s3prl_upstream_name hubert \
+    --feature_layer 6 \
+    --feature_num_clusters 500 \
+    --lang en \
+    --teacher_dumpdir teacher_dumpdir \
```
Comment (kan-bayashi) [inline egs2/ljspeech/tts2/run.sh:23]: Maybe this assumed that durations are already extracted somehow (mfa?).
How about adding information about how to run in README or directly adding the processing in this file?

File: espnet2/tts2/fastspeech2/fastspeech2_discrete.py [hunk 8]
```diff
@@ -0,0 +1,863 @@
...
+import torch
+import torch.nn.functional as F
+from typeguard import typechecked
+
+from espnet2.torch_utils.device_funcs import force_gatherable
+from espnet2.torch_utils.initialize import initialize
+from espnet2.tts2.abs_tts2 import AbsTTS2
+from espnet2.tts2.fastspeech2.loss import FastSpeech2LossDiscrete
+from espnet2.tts2.fastspeech2.variance_predictor import VariancePredictor
+from espnet2.tts2.gst.style_encoder import StyleEncoder
```
Comment (kan-bayashi) [inline espnet2/tts2/fastspeech2/fastspeech2_discrete.py:18]:
```suggestion
from espnet2.tts.fastspeech2.variance_predictor import VariancePredictor
from espnet2.tts.gst.style_encoder import StyleEncoder
```

File: espnet2/tts2/fastspeech2/fastspeech2_discrete.py [hunk 9]
```diff
@@ -0,0 +1,863 @@
...
+        lids: Optional[torch.Tensor] = None,
+        joint_training: bool = False,
+    ) -> Tuple[torch.Tensor, Dict[str, torch.Tensor], torch.Tensor]:
+        """Calculate forward propagation.
+
+        Args:
+            text (LongTensor): Batch of padded token ids (B, T_text).
+            text_lengths (LongTensor): Batch of lengths of each input (B,).
+            feats (Tensor): Batch of padded target features (B, T_feats, odim).
+            feats_lengths (LongTensor): Batch of the lengths of each target (B,).
```
Comment (kan-bayashi) [inline espnet2/tts2/fastspeech2/fastspeech2_discrete.py:510]:
```suggestion
            feats_lengths (LongTensor): Batch of the lengths of each target (B,).
            discrete_speech (Tensor): Discrete speech tensor (B, T_token).
            discrete_speech_lengths (LongTensor): Discrete speech length tensor (B,).
```

File: espnet2/tts2/fastspeech2/fastspeech2_discrete.py [hunk 10]
```diff
@@ -0,0 +1,863 @@
...
+    High-Quality End-to-End Text to Speech`_. Instead of quantized pitch and
+    energy, we use token-averaged value introduced in `FastPitch: Parallel
+    Text-to-speech with Pitch Prediction`_.
+
+    .. _`FastSpeech 2: Fast and High-Quality End-to-End Text to Speech`:
+        https://arxiv.org/abs/2006.04558
+    .. _`FastPitch: Parallel Text-to-speech with Pitch Prediction`:
+        https://arxiv.org/abs/2006.06873
+
+    """
```
Comment (kan-bayashi) [inline espnet2/tts2/fastspeech2/fastspeech2_discrete.py:46]: Please describe the diff from the standard TTS fastspeech2.

File: espnet2/tts2/fastspeech2/fastspeech2_discrete.py (hunk 8 above)
Comment (kan-bayashi) [reply espnet2/tts2/fastspeech2/fastspeech2_discrete.py:18]: Maybe the same.

**Review by kan-bayashi — COMMENTED** [review]
Great work!
I added minor comments, please check them.
And if you want, please modify the header to yours :)

File: egs2/ljspeech/tts2/run.sh [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+set -e
+set -u
+set -o pipefail
+
+train_set=tr_no_dev_10min
+train_dev=dev_small
+eval_set=eval1
+
+# put your vocoder file and vocoder config file here
+vocoder_file=vocoder/vocoder.pkl
```
[the remark this answers is not in the document]
Comment (ftshijt) [reply egs2/ljspeech/tts2/run.sh:13]: I think we might missing the PR to yours. Thanks for reminding! @wyh2000 could you please start a PR to Tomoki's https://github.com/kan-bayashi/ParallelWaveGAN with the recipe of training discrete-based vocoder?

**Comment by ftshijt** [comment]
@sw005320 @simpleoier @kan-bayashi Hi guys, many thanks for your review! I've reflected on your comments and also added some unit tests for the functions added for tts2.

We will have three following PRs after this one (would want to keep them separated to avoid a very large PR)
- tacotron2-based discrete tts2 model
- integration test with tacotron2 (we would want to follow the existing TTS setup with tacotron2)
- alignment-free fastspeech2 with VAE

It would be great if you can double-check the existing PR so that we can merge it in the early place~

File: egs2/ljspeech/tts2/run.sh (hunk 11 above)
[the remark this answers is not in the document]
Comment (wyh2000) [reply egs2/ljspeech/tts2/run.sh]: Thanks for reminding. I start a [PR](https://github.com/kan-bayashi/ParallelWaveGAN/pull/427) to https://github.com/kan-bayashi/ParallelWaveGAN with training and inference recipe. The configurations are aligned with discrete TTS implementation here. Please let me know which one is the best way to share pretrained models. Thanks @kan-bayashi a lot.

**Review by simpleoier — APPROVED** [review]

**Comment by sw005320** [comment]
Thanks for your great efforts, @ftshijt!

**Pull request merged**